forgerock-openbanking-uk-aspsp-common/src/main/java/com/forgerock/openbanking/common/services/JwtOverridingService.java

```diff
@@ -0,0 +1,210 @@
+/**
+ * Copyright 2019 ForgeRock AS.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.forgerock.openbanking.common.services;
+
+import com.forgerock.openbanking.am.config.AMOpenBankingConfiguration;
+import com.forgerock.openbanking.jwt.services.CryptoApiClient;
+import com.forgerock.openbanking.model.oidc.AccessTokenResponse;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+
+import java.text.ParseException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+
+/**
+ * JwtOverridingService provides an API that allows a JWT to be parsed and then rewritten and resigned with the
+ * signing key that the ASPSP has been configured with. This is necessary because we allow AM to process the /authorize
+ * requests, but AM, even if configured to sign with the correct OB Signing key, does not not respond with a jwt that
+ * contains the kid for the key, which means it can't be validated against the OB Sandbox Directory's jwks_uri. So we
+ * must validate, parse, then recreate the id_token, signing it with the OB Signing key (or what ever the
+ * deployment's signing key is configured to be), and assigning a correct kid in the jwk header.
+ */
+@Service
+@Slf4j
+public class JwtOverridingService {
+
+    public class AccessTokenReWriteException extends Exception{
```

Comment:
This might be my personal preference, but I think it's worth making this is a first class citizen (it will also make the catch clauses in the client code a bit more concise).

Reply:
Fixed